File: python/Epa2Shp.py
import shapefile
import yaml, os
import epamodule as d

f = open("arguments.yaml", "r")
cfg = yaml.load(f)

arg=[]
for section in cfg:
    arg.append(cfg[section])

inpname= "%s" % arg[0]
Home = os. getcwd ()

newpath = Home + '\\results'
if not os.path.exists(newpath):
    os.makedirs(newpath)

d.LoadInpFile('networks/'+inpname+'.inp')

# Write Junction Shapefile
w = shapefile.Writer(shapefile.POINT)
w.autoBalance = 1
w.field('dc_id','C',254)
w.field('elevation','N',20)
w.field('pattern','C',254)
w.field('demand','N',20,9)

xy=d.getNodeCoordinates()

x=xy[0]
y=xy[1]
x1=xy[2]
x2=xy[4]
y1=xy[3]
y2=xy[5]
vertx=xy[6]
verty=xy[7]
vertxy=[]
vertxyFinal=[]
for i in range(len(vertx)):
    for u in range(len(vertx[i])):
        vertxy.append([float(vertx[i][u]),float(verty[i][u])])
        vertxyFinal.append(vertxy[u])

if d.getNodeJunctionsCount()>0:
    ndBaseD=d.getNodeBaseDemands()
    ndEle=d.getNodeElevations()
    ndBaseD=d.getNodeBaseDemands()
    ndID=d.getNodeNameID()
    ndPatIndex=d.getNodeDemandPatternIndex()
    ndPatID=d.getPatternNameID()
    for i in range(0,d.getNodeJunctionsCount()):
        w.point(float(x[i]), float(y[i]))
        w.record(ndID[i],ndEle[i],ndPatID[int(ndPatIndex[i]-1)],ndBaseD[i])
    w.save('results/'+inpname[:len(inpname)-4]+'_junctions')

# Write Pipe Shapefile
wpipe = shapefile.Writer(shapefile.POLYLINE)
wpipe.field('dc_id','C',254)
wpipe.field('node1','C',254)
wpipe.field('node2','C',254)
wpipe.field('length','N',20,9)
wpipe.field('diameter','N',20,9)
wpipe.field('status','C',254)
wpipe.field('roughness','N',20,9)
wpipe.field('minorloss','N',20,9)
wpipe.autoBalance = 1
parts=[];pIndex=[]
pumps = d.getLinkPumpIndex()
for i in range(len(pumps)):
    pIndex.append(pumps[i]-1)
vIndex=[]
valves = d.getLinkValveIndex()
for i in range(len(valves)):
    vIndex.append(valves[i]-1)

ndlConn=d.getNodesConnectingLinksID()
if d.getLinkCount()>0:
    p=0;v=0
    linkInStatus=d.getLinkInitialStatus()
    for i in range(0,d.getLinkCount()):
        if i in pIndex:
            xx= (float(x1[i])+float(x2[i]))/2
            yy= (float(y1[i])+float(y2[i]))/2
            for p in range(0,2):
                XY=[]
                if p==0:
                    linkIDFinal=linkID[i]+'_pump1'
                    node1=ndlConn[i][0][0]
                    node2=linkIDFinal
                    indN1 = d.getNodeIndex(node1)
                    XY.append(([float(x[indN1[0]-1]),float(y[indN1[0]-1])],[xx,yy]))
                elif p==1:
                    linkIDFinal=linkID[i]+'_pump2'
                    node1=linkIDFinal
                    node2=ndlConn[i][1][0]
                    indN2 = d.getNodeIndex(node2)

                    XY.append(([xx,yy],[float(x[indN2[0]-1]),float(y[indN2[0]-1])]))
                stat='CLOSED'
                if linkInStatus[i]==1:
                    stat='OPEN'
                length=0
                diameter=0
                roughness=0
                minorloss=0
                wpipe.line(parts=[XY[0]])
                wpipe.record(linkIDFinal,node1,node2,length,diameter,stat,roughness,minorloss)
        elif i in vIndex:
            xx= (float(x1[i])+float(x2[i]))/2
            yy= (float(y1[i])+float(y2[i]))/2
            for v in range(0,2):
                XY=[]
                if v==0:
                    linkIDFinal=linkID[i]+'_valve1'
                    node1=ndlConn[i][0][0]
                    node2=linkIDFinal
                    indN1 = d.getNodeIndex(node1)
                    XY.append(([float(x[indN1[0]-1]),float(y[indN1[0]-1])],[xx,yy]))
                elif v==1:
                    linkIDFinal=linkID[i]+'_valve2'
                    node1=linkIDFinal
                    node2=ndlConn[i][1][0]
                    indN2 = d.getNodeIndex(node2)

                    XY.append(([xx,yy],[float(x[indN2[0]-1]),float(y[indN2[0]-1])]))
                stat='CLOSED'
                if linkInStatus[i]==1:
                    stat='OPEN'
                length=0
                diameter=0
                roughness=0
                minorloss=0
                wpipe.line(parts=[XY[0]])
                wpipe.record(linkIDFinal,node1,node2,length,diameter,stat,roughness,minorloss)
        else:
            linkID=d.getLinkNameID()
            stat='CLOSED'
            if linkInStatus[i]==1:
                stat='OPEN'
            linkLengths=d.getLinkLength()
            linkDiameters=d.getLinkDiameter()
            linkRough=d.getLinkRoughnessCoeff()
            linkMinorloss=d.getLinkMinorLossCoeff()
            point1 = [float(x1[i]),float(y1[i])]
            point2 = [float(x2[i]),float(y2[i])]
            if vertx[i]!=[]:
                parts.append(point1)
                for mm in range(len(vertxyFinal)):
                    parts.append(vertxyFinal[mm])
                parts.append(point2)
                wpipe.line(parts=[parts])
            else:
                wpipe.line(parts=[[[float(x1[i]),float(y1[i])],[float(x2[i]),float(y2[i])]]])
            wpipe.record(linkID[i],ndlConn[i][0][0],ndlConn[i][1][0],linkLengths[i],linkDiameters[i],stat,linkRough[i],linkMinorloss[i])
    wpipe.save('results/'+inpname[:len(inpname)-4]+'_pipes')

# Write Tank Shapefile
w = shapefile.Writer(shapefile.POINT)
w.autoBalance = 1
w.field('dc_id','C',254)
w.field('elevation','N',20)
w.field('initiallev','N',20)
w.field('minimumlev','N',20)
w.field('maximumlev','N',20)
w.field('diameter','N',20)
w.field('minimumvol','N',20)
w.field('volumecurv','N',20)

if d.getNodeTankCount()>0:
    ndTankelevation=d.getNodeElevations()
    initiallev=d.getNodeTankInitialLevel()
    minimumlev=d.getNodeTankMinimumWaterLevel()
    maximumlev=d.getNodeTankMaximumWaterLevel()
    diameter=d.getNodeTankDiameter()
    minimumvol=d.getNodeTankMinimumWaterVolume()
    volumecurv=d.getNodeTankVolumeCurveIndex()
    ndID=d.getNodeNameID()
    for i, nodetankindex in enumerate(d.getNodeTankIndex()):
        p=nodetankindex-1
        w.point(float(x[p]), float(y[p]))
        w.record(ndID[p],ndTankelevation[p],initiallev[p],minimumlev[p],maximumlev[p],diameter[p],minimumvol[p],volumecurv[p])
    w.save('results/'+inpname[:len(inpname)-4]+'_tanks')


# Write Reservoir Shapefile
w = shapefile.Writer(shapefile.POINT)
w.autoBalance = 1
w.field('dc_id','C',254)
w.field('head','N',20)

if d.getNodeReservoirCount()>0:
    head=d.getNodeElevations()
    ndID=d.getNodeNameID()
    for i, nodereservoirindex in enumerate(d.getNodeReservoirIndex()):
        p=nodereservoirindex-1
        w.point(float(x[p]), float(y[p]))
        w.record(ndID[p],head[p])
    w.save('results/'+inpname[:len(inpname)-4]+'_reservoirs')

# Write Pump Shapefile
w = shapefile.Writer(shapefile.POINT)
w.autoBalance = 1
w.field('dc_id','C',254)
w.field('node1','C',254)
w.field('node2','C',254)
w.field('head','C',254)
w.field('flow','C',254)
w.field('power','N',20,9)
w.field('curveID','C',254)


if d.getLinkPumpCount()>0:
    headc=d.getHeadCurveIndex()
    chPowerPump=d.getLinkPumpPower()
    linkID=d.getLinkNameID();power=[]
    for i, pumpindex in enumerate(d.getLinkPumpIndex()):
        Head=[];Flow=[];Curve=[]
        if d.getCurveCount()>0:
            curveXY = d.getCurvesInfo()
            curveXYF = curveXY[headc[i]-1]

            for i in range(len(curveXYF)):
                Head.append(curveXYF[i][2])
                Flow.append(curveXYF[i][3])
            Curve=curveXYF[0][0]
            p=pumpindex-1
            xx= (float(x1[p])+float(x2[p]))/2
            yy= (float(y1[p])+float(y2[p]))/2
            w.point(xx,yy)
            wpipe.line(parts=[[[float(x1[p]),float(y1[p])],[float(x2[p]),float(y2[p])]]])

        else:
            power=chPowerPump[i][1]
            xx= (float(x1[p])+float(x2[p]))/2
            yy= (float(y1[p])+float(y2[p]))/2
            w.point(xx,yy)
            wpipe.line(parts=[[[float(x1[p]),float(y1[p])],[float(x2[p]),float(y2[p])]]])
        if Head==[]:
            Head='NULL'
        if Flow==[]:
            Flow='NULL'
        if Curve==[]:
            Curve='NULL'
        w.record(linkID[p],ndlConn[p][0][0],ndlConn[p][1][0],Head,Flow,power,Curve)
    w.save('results/'+inpname[:len(inpname)-4]+'_pumps')

# Write Valve Shapefile
w = shapefile.Writer(shapefile.POINT)
w.autoBalance = 1
w.field('dc_id','C',254)
w.field('node1','C',254)
w.field('node2','C',254)
w.field('diameter','N',20,9)
w.field('type','C',254)
w.field('setting','N',20,9)
w.field('minorloss','N',20,9)

if d.getLinkValveCount()>0:
    linkID=d.getLinkNameID()
    linkType=d.getLinkType()
    print linkType
    linkDiameter=d.getLinkDiameter()
    linkInitSett=d.getLinkInitialSetting()
    linkMinorloss=d.getLinkMinorLossCoeff()
    for i, valveindex in enumerate(d.getLinkValveIndex()):
        p=valveindex-1
        xx= (float(x1[p])+float(x2[p]))/2
        yy= (float(y1[p])+float(y2[p]))/2
        w.point(xx,yy)
        w.record(linkID[p],ndlConn[p][0][0],ndlConn[p][1][0],linkDiameter[p],linkType[p],linkInitSett[p],linkMinorloss[p])
    w.save('results/'+inpname[:len(inpname)-4]+'_valves')

File: matlab/Epa2Shp.m
function Epa2Shp(inpname)
% which -all mapgate
    format long g;
    d=epanet(inpname);
    warning off;
    Sjunctions=struct;
    Spipes=struct;
    Svalves=struct;
    Spumps=struct;
    Stanks=struct;
    Sreservoirs=struct;
    
    % Write Junction Shapefile
    ndcoords=d.getNodeCoordinates;
    if d.NodeJunctionCount
        for i=1:d.NodeJunctionCount
            Sjunctions(i).('ID')=d.NodeNameID{i};
            Sjunctions(i).('Elevation')=d.NodeElevations(i);
            for u=1:length(d.NodeBaseDemands)
                Sjunctions(i).(['Demand',num2str(u)])=d.NodeBaseDemands{u}(i);
                if ~isempty(d.getPatternNameID{1})
                    if ~isempty(d.NodeDemandPatternNameID{u})
                        Sjunctions(i).(['Pattern',num2str(u)])=d.NodeDemandPatternNameID{u};
                    else
                        Sjunctions(i).(['Pattern',num2str(u)])='None';
                    end
                else
                    Sjunctions(i).(['Pattern',num2str(u)])='None';
                end
            end
            Sjunctions(i).X=ndcoords{1}(i);
            Sjunctions(i).Y=ndcoords{2}(i);
            Sjunctions(i).Geometry='Point';
        end
        [~,f]=fileparts(d.inputfile);
        shapewrite(Sjunctions, ['results/',f,'_junctions.shp']);
    end
    % Write Pipe Shapefile

    if d.LinkCount
        mm=0;qq=0;
        for i=1:d.LinkCount %d.LinkPipeCount
            if sum(i==d.LinkPumpIndex)
                for p=i+mm:i+1+mm
                    if p==i+mm
                        Spipes(p).('ID')=[d.LinkNameID{p-mm},'_pump1'];
                        Spipes(p).('NodeFrom')=d.NodesConnectingLinksID{p-mm,1};
                        Spipes(p).('NodeTo')=Spipes(p).('ID');
                        indN1 = d.getNodeIndex(Spipes(p).('NodeFrom'));
                        indN2 = d.getNodeIndex(d.NodesConnectingLinksID(p-mm,2));

                        Spipes(p).X=[ndcoords{1}(indN1) sum([ndcoords{1}(indN1) ndcoords{1}(indN2)])/2];
                        Spipes(p).Y=[ndcoords{2}(indN1) sum([ndcoords{2}(indN1) ndcoords{2}(indN2)])/2];                        
                        
                    elseif p==i+1+mm
                        Spipes(p).('ID')=[d.LinkNameID{p-mm-1},'_pump2'];
                        Spipes(p).('NodeFrom')=Spipes(p).('ID');
                        Spipes(p).('NodeTo')=d.NodesConnectingLinksID{p-1-mm,2};
                        indN1 = d.getNodeIndex(d.NodesConnectingLinksID(p-1-mm,1));
                        indN2 = d.getNodeIndex(Spipes(p).('NodeTo'));

                        Spipes(p).X=[(sum([ndcoords{1}(indN1) ndcoords{1}(indN2)])/2) ndcoords{1}(indN2)];
                        Spipes(p).Y=[(sum([ndcoords{2}(indN1) ndcoords{2}(indN2)])/2) ndcoords{2}(indN2)];                        
                    end
                    Spipes(p).('Status')='Open';
                    Spipes(p).('Length')=0;
                    Spipes(p).('Diameter')=0;
                    Spipes(p).('Roughness')=0;
                    Spipes(p).('MinorLoss')=0;    
                    Spipes(p).Geometry='Line';
                end
                mm=mm+1;
            elseif sum(i==d.LinkValveIndex)
                for p=i+mm+qq:i+mm+qq+1
                    if p==i+mm+qq
                        Spipes(p).('ID')=[d.LinkNameID{p-mm-qq},'_valve1'];
                        Spipes(p).('NodeFrom')=d.NodesConnectingLinksID{p-mm-qq,1};
                        Spipes(p).('NodeTo')=Spipes(p).('ID');
                        indN1 = d.getNodeIndex(Spipes(p).('NodeFrom'));
                        indN2 = d.getNodeIndex(d.NodesConnectingLinksID(p-mm-qq,2));

                        Spipes(p).X=[ndcoords{1}(indN1) sum([ndcoords{1}(indN1) ndcoords{1}(indN2)])/2];
                        Spipes(p).Y=[ndcoords{2}(indN1) sum([ndcoords{2}(indN1) ndcoords{2}(indN2)])/2];                        
                        
                    elseif p==i+mm+1+qq
                        Spipes(p).('ID')=[d.LinkNameID{p-mm-qq-1},'_valve2'];
                        Spipes(p).('NodeFrom')=Spipes(p).('ID');
                        Spipes(p).('NodeTo')=d.NodesConnectingLinksID{p-mm-qq-1,2};
                        indN1 = d.getNodeIndex(d.NodesConnectingLinksID(p-mm-qq-1,1));
                        indN2 = d.getNodeIndex(Spipes(p).('NodeTo'));

                        Spipes(p).X=[(sum([ndcoords{1}(indN1) ndcoords{1}(indN2)])/2) ndcoords{1}(indN2)];
                        Spipes(p).Y=[(sum([ndcoords{2}(indN1) ndcoords{2}(indN2)])/2) ndcoords{2}(indN2)];                        
                    end
                    Spipes(p).('Status')='Open';
                    Spipes(p).('Length')=0;
                    Spipes(p).('Diameter')=0;
                    Spipes(p).('Roughness')=0;
                    Spipes(p).('MinorLoss')=0;    
                    Spipes(p).Geometry='Line';
                end
                qq=qq+1;
            else
                Spipes(i).('ID')=d.LinkNameID{i};
                Spipes(i).('NodeFrom')=d.NodesConnectingLinksID{i,1};
                Spipes(i).('NodeTo')=d.NodesConnectingLinksID{i,2};
                if d.LinkInitialStatus(i)==1
                    Spipes(i).('Status')='Open';
                else
                    Spipes(i).('Status')='Closed';
                end
                Spipes(i).('Length')=d.LinkLength(i);
                Spipes(i).('Diameter')=d.LinkDiameter(i);
                Spipes(i).('Roughness')=d.LinkRoughnessCoeff(i);
                Spipes(i).('MinorLoss')=d.LinkMinorLossCoeff(i);

                indN1 = d.getNodeIndex(Spipes(i).('NodeFrom'));
                indN2 = d.getNodeIndex(Spipes(i).('NodeTo'));

                % Coordinates for pipes
                Spipes(i).X=[ndcoords{1}(indN1) ndcoords{3}{i} ndcoords{1}(indN2) ];
                Spipes(i).Y=[ndcoords{2}(indN1) ndcoords{4}{i} ndcoords{2}(indN2) ];
                Spipes(i).Geometry='Line';
            end
        end
        [~,f]=fileparts(d.inputfile);
        shapewrite(Spipes, ['results/',f,'_pipes.shp']);
    end
    % Write Tank Shapefile

    if d.NodeTankCount
        u=1;
        for i=d.getNodeTankIndex
            Stanks(u).('ID')=d.NodeNameID{i};        
            Stanks(u).('Elevation')=d.NodeElevations(i);
            Stanks(u).('InitLevel')=d.NodeTankInitialLevel(i);
            Stanks(u).('MinLevel')=d.NodeTankMinimumWaterLevel(i);
            Stanks(u).('MaxLevel')=d.NodeTankMaximumWaterLevel(i);
            Stanks(u).('Diameter')=d.NodeTankDiameter(i);
            Stanks(u).('MinVolume')=d.NodeTankMinimumWaterVolume(i);
            Stanks(u).('VolumeCurve')=d.NodeTankVolumeCurveIndex(i);

            % Coordinates for tanks
            Stanks(u).X=ndcoords{1}(i);
            Stanks(u).Y=ndcoords{2}(i);

            Stanks(u).Geometry='Point';
            u=u+1;
        end
        [~,f]=fileparts(d.inputfile);
        shapewrite(Stanks, ['results/',f,'_tanks.shp']);
    end
    % Write Reservoir Shapefile
        
    if d.NodeReservoirCount
        u=1;
        for i=d.getNodeReservoirIndex
            Sreservoirs(u).('ID')=d.NodeNameID{i};        
            Sreservoirs(u).('Head')=d.NodeElevations(i);

            % Coordinates for reservoirs
            Sreservoirs(u).X=ndcoords{1}(i);
            Sreservoirs(u).Y=ndcoords{2}(i);

            Sreservoirs(u).Geometry='Point';
            u=u+1;
        end
        [~,f]=fileparts(d.inputfile);
        shapewrite(Sreservoirs, ['results/',f,'_reservoirs.shp']);
    end
    
    % Write Pump Shapefile
    if d.LinkPumpCount
        u=1;
        ch=0;
        for i=d.getLinkPumpIndex
            Head='';Flow=''; 
            Spumps(u).('ID')=d.LinkNameID{i};        
            Spumps(u).('NodeFrom')=d.NodesConnectingLinksID{i,1};
            Spumps(u).('NodeTo')=d.NodesConnectingLinksID{i,2};
            headIndex = d.getHeadCurveIndex;
            if sum(headIndex)==0
                Spumps(u).Head=Head;
                Spumps(u).Flow=Flow;
                Spumps(u).Curve='';
                if ch==0
                    linksInfo=d.getBinLinksInfo;
                    ch=1;
                end
                Spumps(u).Power=linksInfo.BinLinkPumpPower(u);
            else
                curveXY = d.getCurveXY(headIndex(u));

                for p=1:length(curveXY(:,1))
                    Head = curveXY(p,1);
                    Flow = curveXY(p,2);
                    Spumps(u).(['Head',num2str(p)])=Head;
                    Spumps(u).(['Flow',num2str(p)])=Flow;
                end
                Spumps(u).Power=0;
                Spumps(u).Curve=d.getCurveNameID(headIndex(u));
            end
            indN1 = d.getNodeIndex(Spumps(u).('NodeFrom'));
            indN2 = d.getNodeIndex(Spumps(u).('NodeTo'));

            % Coordinates for pipes
            Spumps(u).X=sum([ndcoords{1}(indN1) ndcoords{3}{i} ndcoords{1}(indN2) ])/2;
            Spumps(u).Y=sum([ndcoords{2}(indN1) ndcoords{4}{i} ndcoords{2}(indN2) ])/2;

            Spumps(u).Geometry='Point';
            u=u+1;
        end
        [~,f]=fileparts(d.inputfile);
        shapewrite(Spumps, ['results/',f,'_pumps.shp']);
    end
    % Write Valve Shapefile

    if d.LinkValveCount
        u=1;
        for i=d.getLinkValveIndex
            Svalves(u).('ID')=d.LinkNameID{i};        
            Svalves(u).('NodeFrom')=d.NodesConnectingLinksID{i,1};
            Svalves(u).('NodeTo')=d.NodesConnectingLinksID{i,2};

            Svalves(u).('Diameter')=d.LinkDiameter(i);
            Svalves(u).('Type')=d.LinkType{i};
            Svalves(u).('Setting')=d.LinkInitialSetting(i);
            Svalves(u).('MinorLoss')=d.LinkMinorLossCoeff(i);

            indN1 = d.getNodeIndex(Svalves(u).('NodeFrom'));
            indN2 = d.getNodeIndex(Svalves(u).('NodeTo'));

            % Coordinates for pipes
            Svalves(u).X=sum([ndcoords{1}(indN1) ndcoords{3}{i} ndcoords{1}(indN2) ])/2;
            Svalves(u).Y=sum([ndcoords{2}(indN1) ndcoords{4}{i} ndcoords{2}(indN2) ])/2;

            Svalves(u).Geometry='Point';
            u=u+1;
        end
        [~,f]=fileparts(d.inputfile);
        shapewrite(Svalves, ['results/',f,'_valves.shp']);
    end
    d.unload;
    warning on;
